[PointerEvents] pressure property is not spec-compliant when constructed from mouse event

aprotyas opened the pull request:
#### a13dcf305bd1d474c02191d1e4c3c9a517bb6603
<pre>
[PointerEvents] pressure property is not spec-compliant when constructed from mouse event
<a href="https://bugs.webkit.org/show_bug.cgi?id=262159">https://bugs.webkit.org/show_bug.cgi?id=262159</a>
rdar://116094696

Reviewed by Tim Nguyen.

Every PointerEvent object created from a backing MouseEvent object in
PointerCaptureController has an incorrect pressure value, namely 0. This
is the result of the coming together of a few issues:

1. We simply don&apos;t consult MouseEvent.force to initialize
   PointerEvent.pressure when creating a PointerEvent object from a
   backing MouseEvent. This is not the case when constructing from a
   PlatformTouchEvent, where we do correctly consult the force at given
   touch indices.
2. When creating a PointerEvent in a pressure insensitive context, we
   default to initializing the pressure property as zero, which is not
   compliant with the PointerEvents spec.
3. We don&apos;t populate the pressure attribute of NSEvent instances sent by
   EventSenderProxy, so even if we had consulted MouseEvent.force in
   the first issue, we would still be getting the wrong pressure values.

To address these issues, we define a helper static method
`PointerEvent::pressureForPressureInsensitiveInputDevices` that returns
the pressure value defined in the PointerEvents spec, conditioned on the
button currently being pressed. This method is then called if we are in
a pressure insensitive context (i.e. we&apos;re generating a PointerEvent
with Mouse pointer type, rather than Pen/Touch). However, if we&apos;re in a
pressure sensitive context, we simply consult the pressure attribute of
the backing event. Finally, to close the loop and plumb pressure values
from WKTR to generated pointer events, we make sure to supply the
appropriate pressure argument to NSEvent initializers for mouseDown
events.

This patch allows us to pass the following pressure-related assertion in
the pointerevents/pointerevent_attributes_hoverable_pointers.html?mouse WPT.

```
FAIL mouse pointerdown.pressure value is valid assert_greater_than: pressure is greater than 0 with a button pressed expected a number greater than 0 but got 0
```

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_attributes_hoverable_pointers_mouse-expected.txt:

Update test expectations with newly passing assertion.

* Source/WebCore/dom/PointerEvent.cpp:
* Source/WebCore/dom/PointerEvent.h:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent):
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(-[EventSendingController mouseDown:withModifiers:]):
(-[EventSendingController contextClick]):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::EventSenderProxy::mouseDown):
(WTR::EventSenderProxy::sendMouseDownToStartPressureEvents):

Canonical link: <a href="https://commits.webkit.org/268552@main">https://commits.webkit.org/268552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/561945355c7d70dd670ed7231840db9257b01fb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21026 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21874 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18662 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20212 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20556 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20199 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20153 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22726 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17323 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24431 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18400 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22420 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18935 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16061 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18116 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4794 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22465 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18759 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/PointerEvent.h:
```diff
@@ -130,6 +130,13 @@ class PointerEvent final : public MouseEvent {
+    {
+        // https://www.w3.org/TR/pointerevents/#dfn-active-buttons-state
+        bool isInActiveButtonsState = buttons;
+        // https://www.w3.org/TR/pointerevents/#dom-pointerevent-pressure
```
This bit doesn't seem implemented: `Note: all pointerup events will have pressure 0.`

Not sure if there's a test for this somewhere.

aprotyas replied:
This is implemented, `buttons == 0` for all pointerup events, so we return `0` pressure.

aprotyas replied:
The `pointerevents/pointerevent_attributes_hoverable_pointers.html?mouse` WPT makes an assertion about this too, so I think we have enough test coverage.

```js
if (event.buttons === 0) {
    assert_equals(event.pressure, 0, "pressure is 0 with no buttons pressed");
```

nt1m commented on Source/WebKit/Shared/WebEventConversion.cpp:
```diff
@@ -81,7 +81,7 @@ class WebKit2PlatformMouseEvent : public WebCore::PlatformMouseEvent {
             break;
         case WebEventType::MouseForceUp:
             m_type = WebCore::PlatformEvent::Type::MouseForceUp;
-            m_force = WebCore::ForceAtForceClick;
```
Surprised we were putting a value of 2 here! Is the `ForceAtForceClick` variable even correct? Why was it set to 2? The spec says the range is [0, 1]

aprotyas replied:
@nt1m this was introduced in [163047@main](https://commits.webkit.org/163047@main), before the PointerEvents spec was authored, so it makes sense for these values to not match the spec. `ForceAtClick` conveniently aligns with the pressure ranged defined in the specification and has a meaningful name, so I'm reusing it in this context.

nt1m replied:
Should we remove `ForceAtForceClick` ?

aprotyas replied:
As we discussed offline, `ForceAtForceClick` is consequential in some downstream touch event handling code. Having said that, it _is_ a bit confusing since it's not a valid `PointerEvent.pressure` value, so instantiating a `PointerEvent` from a `MouseForceDown` event can put is in a state that does not conform to the PointerEvent spec.

To address this potential error while keeping the touch codepath happy, we choose to clamp incoming pressure values in `PointerEvent` instantiation.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/62ccdbdccfcf91cd3e56ea972a13f86c8e0eb1a2

aprotyas committed:
[PointerEvents] pressure property is not spec-compliant when constructed from mouse event
https://bugs.webkit.org/show_bug.cgi?id=262159
rdar://116094696

Reviewed by Tim Nguyen.

Every PointerEvent object created from a backing MouseEvent object in
PointerCaptureController has an incorrect pressure value, namely 0. This
is the result of the coming together of a few issues:

1. We simply don't consult MouseEvent.force to initialize
   PointerEvent.pressure when creating a PointerEvent object from a
   backing MouseEvent. This is not the case when constructing from a
   PlatformTouchEvent, where we do correctly consult the force at given
   touch indices.
2. When creating a PointerEvent in a pressure insensitive context, we
   default to initializing the pressure property as zero, which is not
   compliant with the PointerEvents spec.
3. We don't populate the pressure attribute of NSEvent instances sent by
   EventSenderProxy, so even if we had consulted MouseEvent.force in
   the first issue, we would still be getting the wrong pressure values.

To address these issues, we define a helper static method
`PointerEvent::pressureForPressureInsensitiveInputDevices` that returns
the pressure value defined in the PointerEvents spec, conditioned on the
button currently being pressed. This method is then called if we are in
a pressure insensitive context (i.e. we're generating a PointerEvent
with Mouse pointer type, rather than Pen/Touch). However, if we're in a
pressure sensitive context, we simply consult the pressure attribute of
the backing event. Finally, to close the loop and plumb pressure values
from WKTR to generated pointer events, we make sure to supply the
appropriate pressure argument to NSEvent initializers for mouseDown
events.

This patch allows us to pass the following pressure-related assertion in
the pointerevents/pointerevent_attributes_hoverable_pointers.html?mouse WPT.

```
FAIL mouse pointerdown.pressure value is valid assert_greater_than: pressure is greater than 0 with a button pressed expected a number greater than 0 but got 0
```

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_attributes_hoverable_pointers_mouse-expected.txt:

Update test expectations with newly passing assertion.

* Source/WebCore/dom/PointerEvent.cpp:
* Source/WebCore/dom/PointerEvent.h:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent):
* Tools/DumpRenderTree/mac/EventSendingController.mm:
(-[EventSendingController mouseDown:withModifiers:]):
(-[EventSendingController contextClick]):
* Tools/WebKitTestRunner/mac/EventSenderProxy.mm:
(WTR::EventSenderProxy::mouseDown):
(WTR::EventSenderProxy::sendMouseDownToStartPressureEvents):

Canonical link: https://commits.webkit.org/268552@main
